src/CodeRunner.ts

```diff
@@ -0,0 +1,277 @@
+     * @param {ProgrammingLanguage} programmingLanguage The language to use
+     */
+    constructor(programmingLanguage: ProgrammingLanguage) {
+        this.programmingLanguage = programmingLanguage;
+        this.editor = new CodeEditor();
+        this.inputManager = new InputManager(async (input: string) => {
+            (await this.backend).writeMessage(input);
+            this.setState(RunState.Running);
+        });
+        this.backend = Promise.resolve({} as SyncClient<Backend>);
+        this.buttons = [];
+        this.addButton({
+            id: RUN_BTN_ID,
+            buttonText: t("Papyros.run"),
+            extraClasses: "text-white bg-blue-500"
+        }, () => this.runCode());
+        this.addButton({
+            id: STOP_BTN_ID,
+            buttonText: t("Papyros.stop"),
+            extraClasses: "text-white bg-red-500"
+        }, () => this.stop());
+        BackendManager.subscribe(BackendEventType.Input,
+            () => this.setState(RunState.AwaitingInput));
+        this.state = RunState.Ready;
+    }
+
+    /**
+     * Start the backend to enable running code
+     */
+    async start(): Promise<void> {
+        this.setState(RunState.Loading);
+        const backend = BackendManager.startBackend(this.programmingLanguage);
+        this.editor.setLanguage(this.programmingLanguage);
+        // Use a Promise to immediately enable running while downloading
+        this.backend = new Promise(resolve => {
+            return backend.workerProxy
+                // Allow passing messages between worker and main thread
+                .launch(proxy((e: BackendEvent) => BackendManager.publish(e)))
+                .then(() => {
+                    this.editor.setCompletionSource(async context => {
+                        const completionContext = Backend.convertCompletionContext(context);
+                        return backend.workerProxy.autocomplete(completionContext);
+                    });
+                    return resolve(backend);
+                });
+        });
+        this.editor.focus();
+        this.setState(RunState.Ready);
+    }
+
+    /**
+     * Interrupt the currently running code
+     * @return {Promise<void>} Promise of stopping
+     */
+    async stop(): Promise<void> {
+        this.setState(RunState.Stopping);
+        BackendManager.publish({
+            type: BackendEventType.End,
+            data: "User cancelled run", contentType: "text/plain"
+        });
+        await this.backend.then(b => b.interrupt());
+    }
+
+    /**
+     * Set the used programming language to the given one to allow editing and running code
+     * @param {ProgrammingLanguage} programmingLanguage The language to use
+     */
+    async setProgrammingLanguage(programmingLanguage: ProgrammingLanguage): Promise<void> {
+        if (this.programmingLanguage !== programmingLanguage) { // Expensive, so ensure it is needed
+            await this.backend.then(b => b.interrupt());
```

Comment:
If you want to be sure here then use client.terminate(), but then the client must be deleted and not reused.